tracer/src/Datadog.Trace/Propagators/W3CContextPropagator.cs

```diff
@@ -0,0 +1,111 @@
+    {
+        /// <summary>
+        /// W3C TraceParent header
+        /// </summary>
+        public const string TraceParent = "traceparent";
+
+        public void Inject<TCarrier, TCarrierSetter>(SpanContext context, TCarrier carrier, TCarrierSetter carrierSetter)
+            where TCarrierSetter : struct, ICarrierSetter<TCarrier>
+        {
+            var traceId = IsValidTraceId(context.RawTraceId) ? context.RawTraceId : context.TraceId.ToString("x32");
+            var spanId = IsValidSpanId(context.RawSpanId) ? context.RawSpanId : context.SpanId.ToString("x16");
+            var sampled = context.SamplingPriority > 0 ? "01" : "00";
+            carrierSetter.Set(carrier, TraceParent, $"00-{traceId}-{spanId}-{sampled}");
+        }
+
+        public bool TryExtract<TCarrier, TCarrierGetter>(TCarrier carrier, TCarrierGetter carrierGetter, out SpanContext? spanContext)
+            where TCarrierGetter : struct, ICarrierGetter<TCarrier>
+        {
+            spanContext = null;
+
+            var traceParent = ParseUtility.ParseString(carrier, carrierGetter, TraceParent)?.Trim();
+            if (!string.IsNullOrEmpty(traceParent))
+            {
+                // We found a trace parent (we are reading from the Http Headers)
+
+                if (traceParent!.Length != 55 || traceParent[2] != '-' || traceParent[35] != '-' || traceParent[52] != '-')
+                {
+                    return false;
+                }
+
+                var w3cSampled = traceParent.Substring(53, 2);
+                if (w3cSampled != "00" && w3cSampled != "01")
+                {
+                    return false;
+                }
+
+                var samplingPriority = w3cSampled == "00" ? 0 : 1;
+
+#if NETCOREAPP
+                var w3cTraceId = traceParent.AsSpan(3, 32);
+                var w3cSpanId = traceParent.AsSpan(36, 16);
+                var traceId = ParseUtility.ParseFromHexOrDefault(w3cTraceId.Slice(16));
+                var parentId = ParseUtility.ParseFromHexOrDefault(w3cSpanId);
+
+                spanContext = new SpanContext(traceId, parentId, samplingPriority, serviceName: null, null)
+                {
+                    RawTraceId = w3cTraceId.ToString(),
+                    RawSpanId = w3cSpanId.ToString()
+                };
+#else
+                var w3cTraceId = traceParent.Substring(3, 32);
+                var w3cSpanId = traceParent.Substring(36, 16);
+                var traceId = ParseUtility.ParseFromHexOrDefault(w3cTraceId.Substring(16));
+                var parentId = ParseUtility.ParseFromHexOrDefault(w3cSpanId);
+
+                spanContext = new SpanContext(traceId, parentId, samplingPriority, serviceName: null, null)
+                {
+                    RawTraceId = w3cTraceId,
+                    RawSpanId = w3cSpanId
+                };
+#endif
+
+                return true;
+            }
+
+            return false;
+        }
+
+        private bool IsValidTraceId(string? traceId)
```

Comment:
```suggestion
        private bool IsValidTraceId([NotNullWhen(true)] string? traceId)
```

Reply:
done in [3bf14ee](https://github.com/DataDog/dd-trace-dotnet/pull/2536/commits/3bf14eec71c141df0e2c99f79cf166748b85b20f)